### ISSUE TITLE
Loading palette not working in PHP 7.1

### ISSUE THREAD

**Issue by ehymel** [szymach/c-pchart#47]
After upgrading to PHP 7.1, the function loadPalette() in Data.php does not correctly load external palettes. This is fixed by forcing (int) type on the numbers read in on line 907:

                $this->Palette[$ID] = ["R" => (int) $R, "G" => (int) $G, "B" => (int) $B, "Alpha" => (int) $Alpha];

without the fix, this results in strings for the values in the Palette array, and in particular the $Alpha is loaded as "100\n", yielding a PHP error "a non well formed numeric value".

**Comment by szymach** [comment]
Hey, can you provide the palette file? I would like to reproduce the error to make sure it is properly resolved by this fix.

**Comment by ehymel** [comment]
This is my palette.txt file (everything between, but not including, the dashed lines):
-------------------------------------------
244, 91,   91, 100
128, 133, 233, 100
141,  70,  84, 100
119, 152, 191, 100
170, 238, 238, 100
255,   0, 102, 100
238, 170, 238, 100
85,  191,  59, 100
223,  83,  83, 100
119, 152, 191, 100
170, 238, 238, 100
-------------------------------------------

**Comment by szymach** [comment]
Hm, by just copy-pasting the contents I can not reproduce the error. Can you send me the file you use? Perhaps is it related to how it is saved. You can just add it in the comment.

**Comment by ehymel** [comment]
[highcharts_palette.txt](https://github.com/szymach/c-pchart/files/1151382/highcharts_palette.txt)

Could be related to how it is saved. The attached file is from a windows system, but the script runs from a linux server.

**Comment by szymach** [comment]
Hey, I have added [test cases](https://github.com/szymach/c-pchart/blob/2.0/tests/unit/ResourceTest.php#L37) for loading your palette file and it seems to be working just fine. [Travis builds](https://travis-ci.org/szymach/c-pchart/builds/254591764) also pass, so I am not sure what more I could do.

Perhaps you can provide a code sample? Or maybe recreate the palette file on Linux and see if the new line symbols were the problem?

**Comment by szymach** [comment]
Hey, any luck with trying a file with Unix-style line endings?

**Comment by ehymel** [comment]
Sorry, I don't have unit testing set up on my simple project. The only thing that fixes the problem on my end is the solution I initially posted, which is to force the loadPalette function to import values at type int; otherwise they are pulled in as strings, including surrounding spaces and the new line symbol. Forcing an (int) type on each imported string solves the problem entirely.

line 907 of Data.php:
`                $this->Palette[$ID] = ["R" => (int) $R, "G" => (int) $G, "B" => (int) $B, "Alpha" => (int) $Alpha];`

**Comment by szymach** [comment]
Hello, sorry for the long delay, had to take a break from coding. Anyways, could you please try this code:

```php
while (!feof($fileHandle)) {
    $line = fgets($fileHandle, 4096);
    if (false === $line) {
        continue;
    }
    $row = explode(',', $line);
    if (empty($row)) {
        continue;
    }
    if (count($row) !== 4) {
        throw new \RuntimeException(sprintf(
            'A palette row must supply R, G, B and Alpha components, %s given!',
            var_export($row, true)
        ));
    }
    list($R, $G, $B, $Alpha) = $row;
    $ID = count($this->Palette);
    $this->Palette[$ID] = [
        "R" => trim($R),
        "G" => trim($G),
        "B" => trim($B),
        "Alpha" => trim($Alpha)
    ];
}
```

and tell me if it changes anything? Thanks.

**Comment by ehymel** [comment]
This appears to work! Thanks for working on it further.

**Comment by ywarnier** [comment]
I was about to report the same issue, but the trim does work, indeed. Could you send this as a new release?
We're using your version of the pChart lib in Chamilo and would love to have that issue fixed (it's nothing really bad as it "only" throws a few dozen "Notice"-level alerts, but it would be nice to have it covered).

https://github.com/chamilo/chamilo-lms/blob/79223b6dda925738ff90c1ddc7b11ebc53fb75db/composer.json#L84

Thank you for making efforts to support 7.1, by the way (looks like 3.0 is all about that).

**Comment by szymach** [comment]
@ywarnier hey, I was waiting for a reply to #48 and was hoping to do both as a single release, but since a week has already passed, I will go ahead and make one today or tomorrow. 

As for the fix itself, it may have been related to `preg_split` (I am not exactly sure if the resulting string was not malformed somehow), but `trim` should also help. 

As for the `3.0` version, I reorganized and simplified the directory structure and got rid of the "factory" class. It was something I created without really knowing much about design patterns all that much and it was pretty hacky and confusing to people. I added PHP 5.4 -> 7.1 support back in the 2.x branch, by the way.
